grpcutil/dial.go

```diff
@@ -94,3 +87,24 @@ func (cred tokenCredProvider) GetRequestMetadata(ctx context.Context, uri ...str
 func (cred tokenCredProvider) RequireTransportSecurity() bool {
 	return false // Allow these to be provided over an insecure channel; this facilitates e.g. service meshes like Istio.
 }
+
+// unaryCompressionInterceptor compresses all outgoing RPCs unless SkipCompressionKey is set.
+func unaryCompressionInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	return invoker(ctx, method, req, reply, cc, compressionInterceptor(ctx, opts)...)
+}
+
+// unaryCompressionInterceptor compresses all outgoing RPCs unless SkipCompressionKey is set.
+func streamCompressionInterceptor(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return streamer(ctx, desc, cc, method, compressionInterceptor(ctx, opts)...)
+}
+
+func compressionInterceptor(ctx context.Context, opts []grpc.CallOption) []grpc.CallOption {
+	if md, ok := metadata.FromIncomingContext(ctx); !ok || len(md.Get(SkipCompressionKey)) == 0 {
+		return append(opts, grpc.UseCompressor(gzip.Name))
```

Comment:
Does the client have to have similar smarts to know if it's compressed or not or does it just look at the message and figure it out?

Reply:
The client makes the decision in download.go (in shouldCompress) based on the filename.  Could potentially get a bit smarter with that but this seems to work well enough for now